test/provider_os_memory_multiple_numa_nodes.cpp

```diff
@@ -111,3 +134,234 @@ TEST_P(testNumaNodes, checkNumaNodesAllocations) {
     int retrieved_numa_node_number = retrieve_numa_node_number(ptr);
     ASSERT_EQ(retrieved_numa_node_number, numa_node_number);
 }
+
+// Test for allocations on numa nodes with mode preferred. It runs for all available
+// numa nodes obtained from the get_available_numa_nodes_numbers() function.
+TEST_P(testNumaOnAllNodes, checkModePreferred) {
+    int numa_node_number = GetParam();
+    umf_os_memory_provider_params_t os_memory_provider_params =
+        UMF_OS_MEMORY_PROVIDER_PARAMS_TEST;
+    os_memory_provider_params.maxnode = numa_node_number + 1;
+    numa_bitmask_setbit(nodemask, numa_node_number);
+    os_memory_provider_params.nodemask = nodemask->maskp;
+    os_memory_provider_params.numa_mode = UMF_NUMA_MODE_PREFERRED;
+    initOsProvider(os_memory_provider_params);
+
+    umf_result_t umf_result;
+    umf_result =
+        umfMemoryProviderAlloc(os_memory_provider, alloc_size, 0, &ptr);
+    ASSERT_EQ(umf_result, UMF_RESULT_SUCCESS);
+    ASSERT_NE(ptr, nullptr);
+
+    // This pointer must point to an initialized value before retrieving a number of
+    // the numa node that the pointer was allocated on (calling get_mempolicy).
+    memset(ptr, 0xFF, alloc_size);
+    int retrieved_numa_node_number = retrieve_numa_node_number(ptr);
+    ASSERT_EQ(retrieved_numa_node_number, numa_node_number);
+}
+
+// Test for allocation on numa node with mode preferred and an empty nodeset.
+// For the empty nodeset the memory is allocated on the node of the CPU that
+// triggered the allocation.
+TEST_F(testNuma, checkModePreferredEmptyNodeset) {
+    umf_os_memory_provider_params_t os_memory_provider_params =
+        UMF_OS_MEMORY_PROVIDER_PARAMS_TEST;
+    os_memory_provider_params.numa_mode = UMF_NUMA_MODE_PREFERRED;
+    initOsProvider(os_memory_provider_params);
+
+    umf_result_t umf_result;
+    umf_result =
+        umfMemoryProviderAlloc(os_memory_provider, alloc_size, 0, &ptr);
+    ASSERT_EQ(umf_result, UMF_RESULT_SUCCESS);
+    ASSERT_NE(ptr, nullptr);
+
+    int cpu = sched_getcpu();
```

Comment:
I think that this test shouldn't be run on a randomized cpu. We don't have a guarantee that this test will be run on different numa nodes.

Maybe you could try parameterizing this test with all available cpus on which the process can be ran and using `sched_setaffinity()` or sth similar to run this test on each of available cpus. This way this test will cover all possible numa nodes with cpus in a multi-numa-node server.